File: src/watermark/site/document_id.py
"""Stable document handles (#1887) — the Python half of the routing-key axis.

A source document is addressed by an opaque 8-character handle derived from its
``data/documents`` rel, not by its path. The as-received path stays untouched on disk and stays
on the page as the custody record; this module only mints the address.

This is a **transcription** of ``web/packages/core/src/documentId.ts``, not an independent
implementation — see that module for why the handle is derived rather than assigned and why the
hash is FNV-1a rather than SHA-256 (the short version: the Cloudflare Pages redirect cap makes a
lookup table impossible, and the same handle has to be computable in Node, the Workers runtime,
and here). Any change to one side must be made to the other in the same commit.

The two are held together by the golden vectors in
``web/packages/core/src/__fixtures__/document-id-vectors.json``, asserted from
``tests/test_site_document_id.py`` and from ``documentId.test.ts``. A drift between the runtimes
would not raise — it would silently 404 every citation — so the vectors are the guard.

Python needs this because retrieval cites documents: the MCP ``search_passages`` tool and the
``/ask`` index emit a document's address alongside its passage.
"""

from __future__ import annotations

from typing import Final

# FNV-1a, 64-bit.
_FNV_OFFSET_BASIS: Final = 0xCBF29CE484222325
_FNV_PRIME: Final = 0x100000001B3
_MASK_64: Final = 0xFFFFFFFFFFFFFFFF

# Crockford base32, lower-cased for URLs — no i/l/o/u, so a handle read aloud or re-typed from a
# filing can't collide through 1/l or 0/O. These handles end up in citations that outlive the build.
_ALPHABET: Final = "0123456789abcdefghjkmnpqrstvwxyz"

#: Handle width in characters: 8 x 5 bits = the 40 bits taken off the hash.
DOCUMENT_ID_LENGTH: Final = 8

#: Bits discarded from the 64-bit hash to leave a whole number of base32 characters.
_DISCARDED_BITS: Final = 64 - DOCUMENT_ID_LENGTH * 5

#: Curated handle pins — the escape hatch for a document that **moves** to a different rel and
#: must keep the handle it was already cited under. Must stay byte-identical to
#: ``DOCUMENT_ID_PINS`` in ``documentId.ts``; the vector fixture covers the derivation, this dict
#: is covered by ``test_pins_match_frontend``.
DOCUMENT_ID_PINS: Final[dict[str, str]] = {}


def _fnv1a64(value: str) -> int:
    """FNV-1a over the UTF-8 bytes of ``value``."""
    digest = _FNV_OFFSET_BASIS
    for byte in value.encode("utf-8"):
        digest = ((digest ^ byte) * _FNV_PRIME) & _MASK_64
    return digest


def _encode_base32(value: int) -> str:
    """Big-endian Crockford base32 of the low ``DOCUMENT_ID_LENGTH * 5`` bits of ``value``."""
    out: list[str] = []
    remaining = value
    for _ in range(DOCUMENT_ID_LENGTH):
        out.append(_ALPHABET[remaining & 31])
        remaining >>= 5
    return "".join(reversed(out))


def document_id(rel: str) -> str:
    """The stable handle for a ``data/documents`` rel.

    Takes the rel **verbatim**, exactly as it appears in the documents feed: no case folding, no
    separator normalization, no URL decoding. The rel is the as-received chain-of-custody path,
    and normalizing here would be a second, silently-diverging definition of it.
    """
    pinned = DOCUMENT_ID_PINS.get(rel)
    if pinned is not None:
        return pinned
    return _encode_base32(_fnv1a64(rel) >> _DISCARDED_BITS)


def doc_permalink(document: str) -> str:
    """The site-relative permalink for a handle (``/doc/<id>/``), carrying no site base."""
    return f"/doc/{document}/"


def doc_permalink_for_rel(rel: str) -> str:
    """``doc_permalink(document_id(rel))`` — the common one-step call."""
    return doc_permalink(document_id(rel))

File: web/packages/core/src/documentId.ts
/**
 * Stable document handles (#1887, epic #1884 phase 3) — the **routing** key for a source document.
 *
 * The corpus's on-disk tree is not the URL tree. A document is addressed by an opaque 8-character
 * handle derived from its `data/documents` rel; the as-received path stays untouched on disk and
 * stays on the page as the custody record. This module is the whole ID axis: no other file may
 * mint one.
 *
 * Why derived rather than assigned, and why a non-crypto hash:
 *
 *  - **The redirect budget.** Cloudflare Pages caps `_redirects` at 2,000 static + 100 dynamic
 *    rules; Lima alone has 3,247 documents whose deep legacy paths must keep resolving. Because
 *    the handle is a pure function of the rel, the legacy-path Function computes it at the edge
 *    and 301s with no lookup table — one file instead of 3,247 impossible redirect lines.
 *  - **Three runtimes, one answer.** Node (the Astro build), the Workers runtime (the redirect
 *    Function and `/api/ask` citation rendering), and Python (`watermark.site.document_id`, for
 *    the MCP `search_passages` citation) must agree byte-for-byte or a citation silently 404s.
 *    `crypto.subtle.digest` is async and would poison every sync call site, so this is FNV-1a —
 *    trivially portable, and collision resistance is not a security property here. Integrity is
 *    already carried by the Git-LFS oid / R2 etag; this is a routing key. `documentId.test.ts`
 *    asserts zero collisions across the committed corpus, and the golden vectors in
 *    `__fixtures__/document-id-vectors.json` are asserted from both Node and pytest.
 *  - **Stable across the re-pull that actually happens.** A source re-served at the *same* rel
 *    with new bytes (the Ohio EPA DAM re-serving `permits/doc/` in place) keeps its handle, which
 *    is correct — it is the same record. A byte change at a stable rel is a *version* event, and
 *    `data/site/document-versions.yaml` (#1590) already carries that vocabulary.
 *
 * The handle deliberately does not encode the filename: that is what produced the 16-segment
 * routes this replaces.
 */

// FNV-1a, 64-bit. BigInt keeps the arithmetic exact and the port to Python a transcription.
const FNV_OFFSET_BASIS = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const MASK_64 = 0xffffffffffffffffn;

/**
 * Crockford base32, lower-cased for URLs — no `i`/`l`/`o`/`u`, so a handle read aloud, written
 * down, or re-typed from a filing can't collide through `1`/`l` or `0`/`O`. These handles end up
 * in citations that outlive the build, so the alphabet matters.
 */
const ALPHABET = "0123456789abcdefghjkmnpqrstvwxyz";

/** Handle width in characters: 8 × 5 bits = the 40 bits taken off the hash. */
export const DOCUMENT_ID_LENGTH = 8;

/** Bits discarded from the 64-bit hash to leave a whole number of base32 characters. */
const DISCARDED_BITS = 64n - BigInt(DOCUMENT_ID_LENGTH) * 5n;

const DOCUMENT_ID_RE = new RegExp(`^[${ALPHABET}]{${DOCUMENT_ID_LENGTH}}$`);

/**
 * Curated handle pins — the escape hatch for the one case the derivation can't cover: a document
 * that **moves** to a different rel and must keep the handle it was already cited under.
 *
 * Same shape and same discipline as `HANDLE_RENAMES` (#1099): keep it small, keep it reviewed,
 * and add an entry only with the old and new rels recorded in the commit. Left empty, every
 * handle is reproducible from the corpus alone. Note that old *URLs* keep working without a pin —
 * the legacy path still hashes to the handle it was minted under; a pin is needed only so the
 * document's *current* rel resolves to that same handle.
 */
export const DOCUMENT_ID_PINS: Readonly<Record<string, string>> = {
  // "oepa/van-wert/old-name.pdf": "7k3m9qpb",
};

/** FNV-1a over the UTF-8 bytes of `input`. `TextEncoder` is present in Node, Workers, and browsers. */
function fnv1a64(input: string): bigint {
  let hash = FNV_OFFSET_BASIS;
  for (const byte of new TextEncoder().encode(input)) {
    hash = ((hash ^ BigInt(byte)) * FNV_PRIME) & MASK_64;
  }
  return hash;
}

/** Big-endian Crockford base32 of the low `DOCUMENT_ID_LENGTH * 5` bits of `value`. */
function encodeBase32(value: bigint): string {
  let out = "";
  let remaining = value;
  for (let i = 0; i < DOCUMENT_ID_LENGTH; i++) {
    out = ALPHABET[Number(remaining & 31n)] + out;
    remaining >>= 5n;
  }
  return out;
}

/**
 * The stable handle for a `data/documents` rel — the document's address everywhere in the site.
 *
 * Takes the rel **verbatim**, exactly as it appears in the documents feed: no case folding, no
 * separator normalization, no URL decoding. The rel is the as-received chain-of-custody path and
 * any normalization here would be a second, silently-diverging definition of it.
 */
export function documentId(rel: string): string {
  const pinned = DOCUMENT_ID_PINS[rel];
  if (pinned !== undefined) return pinned;
  return encodeBase32(fnv1a64(rel) >> DISCARDED_BITS);
}

/** Whether `value` is well-formed as a handle — a cheap route-param / redirect-target guard. */
export function isDocumentId(value: string): boolean {
  return DOCUMENT_ID_RE.test(value);
}

/**
 * The site-relative permalink for a handle (`/doc/<id>/`). Callers wrap it in `withSite` /
 * `withBasePath`; it carries no base of its own, matching `docPagePath`'s contract.
 *
 * Flat and collection-free on purpose: the address survives any re-shelving of the corpus, and
 * the collection/container/folder path renders *on* the page as provenance instead.
 */
export function docPermalink(id: string): string {
  return `/doc/${id}/`;
}

/** `docPermalink(documentId(rel))` — the common one-step call. */
export function docPermalinkForRel(rel: string): string {
  return docPermalink(documentId(rel));
}
